added more tests and perc_query_aligned param and blastn-short setting

nbokulich opened the pull request:
fixes #8

gregcaporaso commented on q2_quality_control/_blast.py:
```diff
@@ -72,7 +87,17 @@ def _extract_hits(blast_output):
                 # if vsearch fails to find assignment, it reports '*' as the
                 # accession ID, so we will not count those IDs as hits.
-                and line.split('\t')[1] != '*']
+                and line.split('\t')[1] != '*'
```
`line` is being split four times here, which is adding some unnecessary computation. I recommend making this a `for` loop rather than a list comprehension (which is clearer anyway, where there are a lot of conditions), and splitting only one time (e.g., `fields = line.split('\t')`).

nbokulich replied:
👍

gregcaporaso commented on q2_quality_control/quality_control.py:
```diff
@@ -19,20 +19,28 @@
+        res_md = qiime2.Metadata(res)
+        hits_seqs = filter_seqs(query_series, res_md, exclude_ids=False)
+    except ValueError:
+        raise ValueError('No query sequences match the reference.')
```
Do we really want the method to fail in these cases? Or should we be creating empty objects? We are able to create empty `FeatureData[Sequence]` objects, as this came up when I was implementing `cluster-features-closed-reference` in `q2-vsearch`.

I realize that it's less important in this case than the closed-reference case (since if you filter no sequences you continue with your input table, and if you feature all... well... you have bigger problems), but it could still be annoying, for example if a user is setting up an automated processing pipeline that includes this filtering process.

nbokulich replied:
👍 
Where is this implemented in `cluster-features-closed-reference` though? It looks like a `ValueError` is raised if there are no matches to ref.

nbokulich replied:
Either way I see this conflicting with automated pipelines if empty jobs are output, but I agree that may be less of an issue than just failing outright (failing will always cause failure, empty files will only cause failure if they are empty!)

Should we perhaps raise a warning or something to report seq counts?

gregcaporaso replied:
@nbokulich, I was referring to empty [`unmatched_seqs` in `cluster-features-closed-reference`](https://github.com/qiime2/q2-vsearch/blob/de455f66c4c10a0227f116410ae4dfe78222a37c/q2_vsearch/_cluster_features.py#L170), since that result is the same semantic type as the return types here. (No matches to the reference in that case would be an empty feature table, which is something that I think we don't have support for at the moment, and which it makes a little more sense to fail over. Failing because all sequences match the reference would be kind of a bummer...)

gregcaporaso commented on q2_quality_control/tests/test_quality_control.py:
```diff
@@ -38,6 +38,10 @@ def _load_DNAFASTAFormat(reads_fn):
             reads_fp = self.get_data_path(reads_fn)
             return DNAFASTAFormat(reads_fp, mode='r')
 
+        def _dnafastaformats_to_series(fasta):
```
:+1:

gregcaporaso commented on q2_quality_control/tests/test_quality_control.py:
```diff
@@ -46,19 +50,106 @@ def _load_DNAFASTAFormat(reads_fn):
+            'query-partially-random.fasta')
+
+    def test_exclude_seqs_bacterial_hit_fungal_miss(self):
+        for method in ['blast', 'vsearch']:
```
In these tests where you're iterating over the methods, do you know which method was used when you get a test failure? If not, that could be annoying to debug later on. If the method isn't reported, you could include that in the [optional `msg` parameter to `assertEqual`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual), or I could give you some other suggestions for how to structure this to avoid duplicating the tests (e.g., you could have a base class with the tests, and then override that in subclasses that have a `method` variable that is used in all of the tests).

nbokulich replied:
@gregcaporaso thanks for the suggestion about test subclasses — if it won't take long, could you offer some pseudocode to direct me as I implement this?

nbokulich replied:
👍

nbokulich commented:
Thanks @gregcaporaso ! I have added inheritance to unit test subclasses for blast/vsearch

I have made `_extract_hits` a for loop instead of loop comprehension

I have returned the behavior to output empty files instead of raising an error (I changed to this behavior because I thought this is what `q2-vsearch` did... but I agree we want this streamlined for batch jobs).

gregcaporaso commented on q2_quality_control/quality_control.py:
```diff
@@ -17,21 +19,37 @@
+        res_md = qiime2.Metadata(res)
+        hits_seqs = filter_seqs(query_series, res_md, exclude_ids=False)
+    except ValueError:
+        raise ValueError('No query sequences match the reference.')
```
Is it possible to reach this except block? It looks like this is handled above.

nbokulich replied:
you are correct — now that I check the length of `res` against `query_series` these errors should never be reached.

gregcaporaso commented:
Thanks @nbokulich!
